pkg/services/x509/x509_validator.go

```diff
@@ -0,0 +1,272 @@
+package x509
+
+import (
+	"bytes"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/lestrrat-go/jwx/jwa"
+	"github.com/lestrrat-go/jwx/jws"
+	"github.com/lestrrat-go/jwx/jwt"
+)
+
+type JwtX509Token struct {
+	chain  []*x509.Certificate
+	token  jwt.Token
+	raw    string
+	sigAlg jwa.SignatureAlgorithm
+}
+
+type JwtX509Validator struct {
+	roots              []*x509.Certificate
+	intermediates      []*x509.Certificate
+	allowedSigningAlgs []jwa.SignatureAlgorithm
+}
+
+// generalNames defines the asn1 data structure of the generalNames as defined in rfc5280#section-4.2.1.6
+type generalNames struct {
+	OtherName otherName `asn1:"tag:0"`
+}
+
+// otherName defines the asn1 data structure of the othername as defined in rfc5280#section-4.2.1.6
+type otherName struct {
+	OID   asn1.ObjectIdentifier
+	Value asn1.RawValue `asn1:"tag:0"`
+}
+
+// is the object identifier of the subjectAltName (id-ce 17)
+var subjectAltNameID = asn1.ObjectIdentifier{2, 5, 29, 17}
+
+func (j JwtX509Token) SubjectAltNameOtherName() (string, error) {
+	leaf := j.chain[0]
+	var subjAltName pkix.Extension
+	for _, ext := range leaf.Extensions {
+		if ext.Id.Equal(subjectAltNameID) {
+			subjAltName = ext
+		}
+		break
+	}
+	if len(subjAltName.Value) == 0 {
+		return "", nil
+	}
+
+	var otherNames generalNames
+	if _, err := asn1.Unmarshal(subjAltName.Value, &otherNames); err != nil {
+		return "", err
+	}
+	var otherNameStr string
+	if _, err := asn1.Unmarshal(otherNames.OtherName.Value.Bytes, &otherNameStr); err != nil {
+		return "", err
+	}
+	return otherNameStr, nil
+}
+
+func NewJwtX509Validator(roots, intermediates []*x509.Certificate, allowedSigAlgs []jwa.SignatureAlgorithm) *JwtX509Validator {
+	return &JwtX509Validator{
+		roots:              roots,
+		intermediates:      intermediates,
+		allowedSigningAlgs: allowedSigAlgs,
+	}
+}
+
+// Parse attempts to parse a string as a jws. It checks if the x5c header contains at least 1 certificate.
+// The signature should be signed with the private key of the leaf certificate.
+// No other validations are performed. Call Verify to verify the auth token.
+func (validator JwtX509Validator) Parse(rawAuthToken string) (*JwtX509Token, error) {
+	// check the cert chain in the jwt
+	rawHeader, _, _, err := jws.SplitCompact(bytes.NewReader([]byte(rawAuthToken)))
+	if err != nil {
+		return nil, fmt.Errorf("the jwt should contain of 3 parts: %w", err)
+	}
+
+	headers := jws.NewHeaders()
+	headerStr, _ := base64.RawURLEncoding.DecodeString(string(rawHeader))
+	if err = json.Unmarshal(headerStr, headers); err != nil {
+		return nil, fmt.Errorf("could not unmarshall headers: %w", err)
+	}
+
+	certsFromHeader := headers.X509CertChain()
+	if len(certsFromHeader) == 0 {
+		return nil, fmt.Errorf("the jwt x5c field should contain at least 1 certificate")
```

Comment:
nothing to format

Reply:
I like one way of creating my errors. fmt.Errorf is the most flexible. Under the hood they are the same.

Reply:
I think it's important we settle on a standard which is in line with the rest of the code. Argument in favor of using `errors.New` when no formatting is involved: compiler warn or even fail on missing or superfluous formatting directives. At least for `(S)printf`